rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,395 @@
+This makes it much easier to contribute new packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- It is not obvious to package contributors where to add files or which ones to edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in `all-packages.nix` should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in `all-packages.nix` to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even locks up editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
+These PR's should be done after a release to maximize the testing period and minimize merge conflicts.
+
+## PR 1: The directory structure
+
+This part establishes the new directory structure in Nixpkgs.
+This directory structure is internal to Nixpkgs and not exposed as public interface.
+This directory structure must be documented in the Nixpkgs manual.
+This PR will be backported to the stable release in order to ensure that backports of new packages work.
+
+### File structure
+
+Create the initially-empty `pkgs/by-name` directory in Nixpkgs, and migrate the `hello` package into it.
+
+Check the following using CI:
+- `pkgs/by-name` must only contain subdirectories of the form `${shard}/${name}`, called _package directories_.
+- The `name`'s of package directories must be unique when lowercased
+- `name` is a string only consisting of the ASCII characters `a-z`, `A-Z`, `0-9`, `-` or `_`.
+- `shard` is the lowercased first two letters of `name`, expressed in Nix: `shard = toLower (substring 0 2 name)`.
+- Each package directory must contain a `package.nix` file and may contain arbitrary other files.
+
+### Semantics
+
+Introduce code to automatically define `pkgs.${name}` for each package directory as a value equivalent to
```

Comment:
(obviously a bit late to the party, sorry about that)

wouldn't this code generate a ton of `getdents` syscalls? (and degrades performance of evaluation over time as packages are added)

Reply:
Please have a look at the performance discussion that happened on the implementation PRs.

Reply:
I checked the two other PRs, but didn't check this one. Thanks!

https://github.com/NixOS/nixpkgs/pull/237439#issuecomment-1640686296 (and following comments)